projection/src/test/scala/akka/projection/r2dbc/R2dbcTimestampOffsetStoreSpec.scala

```diff
@@ -212,12 +215,55 @@ class R2dbcTimestampOffsetStoreSpec
       // and subsequent seqNr is accepted
       offsetStore.isSequenceNumberAccepted(createEnvelope("p5", 8L, startTime.plusMillis(9), 10, "e5-8")) shouldBe true
 
-      // it's keeping the seen that are not in the "stored" state
-      offsetStore.getSeen() shouldBe Map("p1" -> 4L, "p3" -> 8L, "p4" -> 2L, "p5" -> 8L)
+      // it's keeping the inflight that are not in the "stored" state
+      offsetStore.getInflight() shouldBe Map(
+        "p1" -> Processing(4L),
+        "p3" -> Processing(8L),
+        "p4" -> Processing(2L),
+        "p5" -> Processing(8L))
       // and they are removed from seen once they have been stored
       offsetStore.saveOffset(TimestampOffset(startTime.plusMillis(2), Map("p4" -> 2L))).futureValue
       offsetStore.saveOffset(TimestampOffset(startTime.plusMillis(9), Map("p5" -> 8L))).futureValue
-      offsetStore.getSeen() shouldBe Map("p1" -> 4L, "p3" -> 8L)
+      offsetStore.getInflight() shouldBe Map("p1" -> Processing(4L), "p3" -> Processing(8L))
+    }
+
+    "update inflight on error and re-accept element" in {
+      val projectionId = genRandomProjectionId()
+      val offsetStore = createOffsetStore(projectionId)
+
+      val startTime = Instant.now()
+
+      val envelope1 = createEnvelope("p1", 1L, startTime.plusMillis(1), 10, "e1-1")
+      val envelope2 = createEnvelope("p1", 2L, startTime.plusMillis(2), 10, "e1-2")
+      val envelope3 = createEnvelope("p1", 3L, startTime.plusMillis(2), 10, "e1-2")
+
+      // seqNr 1 is always accepted
+      offsetStore.isSequenceNumberAccepted(envelope1) shouldBe true
+      offsetStore.getInflight() shouldBe Map("p1" -> Processing(1L))
+      offsetStore.saveOffset(TimestampOffset(startTime.plusMillis(1), Map("p1" -> 1L))).futureValue
+      offsetStore.getInflight() shouldBe empty
+
+      // seqNr 2 is accepts since it follows seqNr 1 that is stored in state
+      offsetStore.isSequenceNumberAccepted(envelope2) shouldBe true
+      // simulate envelope processing error
+      offsetStore.updateInflightOnError(envelope2)
+      offsetStore.getInflight() shouldBe Map("p1" -> Recovering(2L))
+
+      // seqNr 3 is not accepted, still waiting for seqNr 2
+      offsetStore.isSequenceNumberAccepted(envelope3) shouldBe false
+
+      // offer seqNr 2 once again
+      offsetStore.isSequenceNumberAccepted(envelope2) shouldBe true
+      // is back to processing
+      offsetStore.getInflight() shouldBe Map("p1" -> Processing(2L))
+
+      // offer seqNr 3  once more
+      offsetStore.isSequenceNumberAccepted(envelope3) shouldBe true
+      offsetStore.getInflight() shouldBe Map("p1" -> Processing(3L))
```

Comment:
> The only case I could imagine for this to happen, and probably the reason why it's implemented like that, is when using grouped with async (non-transactional).

That is not the only reason. When using any async the offset is saved afterwards and several additional events can be processed before the offset is saved. We have have batching of offset saving after  20 events or 500 millis.

Reply:
@patriknw, I don't know if I follow you now but we can catch-up later.

In any case, my point is that whenever we have any async projection the inflight for a given pid is 1 or more envelopes and this information is not being preserved with the current modeling we have. We only have the latest one. 

I have worked further on the idea of using a `pid -> Inflight(Seq[SeqNr])` to represent it.  So the `Inflight` would hold a seq of numbers for each pid. But I realise that my tentative design is wrong. So, I'm not pushing that code now because I don't want to break this PR. I will back up it on another branch. 

What I have in mind is the following. Given pid = abc and envelopers e1, e2 and e3. 

On first run, we will have: "abc" -> Seq(Processing(e1), Processing(e2), Processing(e3))

If the projection fails, it moves to:
"abc" -> Seq(Recovering(e1), Recovering(e2), Recovering(e3))
Meaning, all envelopes for that pid will need to be reprocessed. 

If it's a groupedWithin, it means that the group will need to be consumed again. 
If it's a at-least-once, we should be able to remove items that have been processed successfully (if possible).  

The envelopes get offered one by one and we move gradually to:
"abc" -> Seq(Processing(e1), Recovering(e2), Recovering(e3))
"abc" -> Seq(Processing(e1), Processing(e2), Recovering(e3))
"abc" -> Seq(Processing(e1), Processing(e2), Processing(e3))

As such we can make the right decision is an envelope can be accepted again. 

Anyway, this is idea still need to be proved correct. I will explore it when I'm back. For the time being this PR can be reviewed and eventually merged. Otherwise, later, this is not that urgent now.

Reply:
I haven't looked at this PR yet and haven't thought so much about how to solve the problem, but my initial thought would be that if processing fails we want to revert to the previous Map(pid -> seqNr).

I think that can be implemented in a rather clean way since we have an immutable Map so we can set it back to previous Map. Alternatively we separate the check from the update and only update after the processing has been successful.